---
title: "Dynamics of within-species Variation"
subtitle: "`r paste0('Genome: ', cur_genome, '. Taxonomy: ', tax_df$Tax)`"
author: "Giulio Benedetti"
date: "`r Sys.Date()`"
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(fig.path = paste0("results/", cur_genome, "_"),
                      echo = FALSE)
```

```{r}
p1 <- ggplot(count_df, aes(x = Time,
                           y = Percent,
                           colour = Season,
                           group = RefSample)) +
  geom_point() +
  geom_line() +
  scale_x_date(date_breaks = "1 month") +
  scale_y_continuous(limits = c(0, 1),
                     breaks = seq(0, 1, by = 0.1)) +
  labs(y = "Percent of Shared Polymorphic Sites",
       colour = "Reference Sample") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        panel.grid = element_blank())
```

```{r}
p3 <- ggplot(genome_time_series_df, aes(x = Sample)) +
  geom_line(aes(y = RPKM), colour = "Black") +
  geom_line(aes(y = NucDiv * 1000), colour = "Dark Gray") +
  scale_x_date(date_breaks = "1 month") +
  scale_y_continuous(name = "RPKM",
                     breaks = seq(0, up_lim, by = 10),
                     sec.axis = sec_axis(~./1000,
                                         name = "Nucleotide Diversity")) +
  theme_bw() +
  labs(x = "Time") +
  theme(axis.title.y = element_text(colour = "Black"),
        axis.title.y.right = element_text(colour = "Dark Gray"),
        panel.grid = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1))
```

```{r polyline, fig.height = 10, fig.width = 20}
p <- (p2 / p3 | p1) +
    plot_layout(guides = "collect")
p
```

```{r polymap, fig.height = 15, fig.width = 15}
draw(Heatmap(count_mat,
             name = "SPS Percent",
             heatmap_legend_param = list(at = seq(0, 1, by = 0.2)),
             col = col_fun,
             cluster_rows = FALSE,
             cluster_columns = FALSE,
             row_title = "Reference Sample",
             column_title = "Observed Sample",
             column_names_rot = 90,
             left_annotation = row_ha,
             bottom_annotation = col_ha_bottom,
             top_annotation = col_ha_top,
             show_row_names = FALSE,
             show_column_names = FALSE),
     annotation_legend_list = list(lgd_gvalue, lgd_sig))
```
